piker/data/_sampling.py

```diff
@@ -0,0 +1,240 @@
+    # to solve this is to make this task aware of the instrument's
+    # tradable hours?
+
+    global _incrementers
+
+    # adjust delay to compensate for trio processing time
+    ad = min(_shms.keys()) - 0.001
+
+    total_s = 0  # total seconds counted
+    lowest = min(_shms.keys())
+    ad = lowest - 0.001
+
+    with trio.CancelScope() as cs:
+
+        # register this time period step as active
+        _incrementers[delay_s] = cs
+        task_status.started(cs)
+
+        while True:
+            # TODO: do we want to support dynamically
+            # adding a "lower" lowest increment period?
+            await trio.sleep(ad)
+            total_s += lowest
+
+            # increment all subscribed shm arrays
+            # TODO: this in ``numba``
+            for delay_s, shms in _shms.items():
+                if total_s % delay_s != 0:
+                    continue
+
+                # TODO: ``numba`` this!
+                for shm in shms:
+                    # TODO: in theory we could make this faster by copying the
+                    # "last" readable value into the underlying larger buffer's
+                    # next value and then incrementing the counter instead of
+                    # using ``.push()``?
+
+                    # append new entry to buffer thus "incrementing" the bar
+                    array = shm.array
+                    last = array[-1:][shm._write_fields].copy()
+                    # (index, t, close) = last[0][['index', 'time', 'close']]
+                    (t, close) = last[0][['time', 'close']]
+
+                    # this copies non-std fields (eg. vwap) from the last datum
+                    last[
+                        ['time', 'volume', 'open', 'high', 'low', 'close']
+                    ][0] = (t + delay_s, 0, close, close, close, close)
+
+                    # write to the buffer
+                    shm.push(last)
+
+                # broadcast the buffer index step
+                # yield {'index': shm._last.value}
+                for ctx in _subscribers.get(delay_s, ()):
+                    try:
+                        await ctx.send_yield({'index': shm._last.value})
+                    except (
+                        trio.BrokenResourceError,
+                        trio.ClosedResourceError
+                    ):
+                        log.error(f'{ctx.chan.uid} dropped connection')
+
+
+@tractor.stream
+async def iter_ohlc_periods(
+    ctx: tractor.Context,
+    delay_s: int,
+) -> None:
+    """
+    Subscribe to OHLC sampling "step" events: when the time
+    aggregation period increments, this event stream emits an index
+    event.
+
+    """
+    # add our subscription
+    global _subscribers
+    subs = _subscribers.setdefault(delay_s, [])
+    subs.append(ctx)
+
+    try:
+        # stream and block until cancelled
+        await trio.sleep_forever()
+    finally:
+        subs.remove(ctx)
+
+
+async def sample_and_broadcast(
```

Comment:
There I think isolating these shm sampling routines is much more easy for code readers.